<launch>
  <set_parameter name="use_sim_time" value="true" />

  <arg name="rviz" default="false" description="launch rviz"/>
  <arg name="map_path" default="" description="point cloud and lanelet2 map directory path"/>
  <arg name="save_dir" default="/home/$(env USER)/deviation_evaluator_sample"/>
  <arg name="stddev_vx" default="0.8" description="[m/s]"/>
  <arg name="stddev_wz" default="0.01" description="[rad/s]"/>
  <arg name="coef_vx" default="1.0" description="[m/s]"/>
  <arg name="bias_wz" default="0.0" description="[]"/>
  <arg name="period" default="10.0" description="[s]"/>
  <arg name="cut" default="9.0" description="[s]"/>

  <arg name="in_twist_with_covariance" default="/localization/twist_estimator/twist_with_covariance"/>
  <arg name="in_ndt_pose_with_covariance" default="/localization/pose_estimator/pose_with_covariance"/>
  <arg name="in_ekf_odom" default="/localization/deviation_evaluator/ekf_localizer/kinematic_state"/>
  <arg name="in_ekf_gt_odom" default="/localization/pose_twist_fusion_filter/kinematic_state"/>
  <arg name="out_twist_with_covariance" default="/localization/deviation_evaluator/twist_estimator/twist_with_covariance"/>
  <arg name="out_ndt_pose_with_covariance" default="/localization/deviation_evaluator/pose_estimator/pose_with_covariance"/>
  <arg name="lanelet2_map_file" default="lanelet2_map.osm" />
  <arg name="pointcloud_map_file" default="pointcloud_map.pcd" />

  <node pkg="deviation_evaluator" exec="deviation_evaluator" name="deviation_evaluator" output="screen">
    <remap from="in_twist_with_covariance" to="$(var in_twist_with_covariance)"/>
    <remap from="in_ndt_pose_with_covariance" to="$(var in_ndt_pose_with_covariance)"/>
    <remap from="out_twist_with_covariance" to="$(var out_twist_with_covariance)"/>
    <remap from="out_ndt_pose_with_covariance" to="$(var out_ndt_pose_with_covariance)"/>

    <param name="save_dir" value="$(var save_dir)"/>
    <param name="stddev_vx" value="$(var stddev_vx)" />
    <param name="stddev_wz" value="$(var stddev_wz)" />
    <param name="coef_vx" value="$(var coef_vx)" />
    <param name="bias_wz" value="$(var bias_wz)" />
    <param name="period" value="$(var period)" />
    <param name="cut" value="$(var cut)" />
  </node>

  <include file="$(find-pkg-share ekf_localizer)/launch/ekf_localizer.launch.xml">
    <arg name="extend_state_step" value="50"/>
    <arg name="enable_yaw_bias_estimation" value="false"/>
    <arg name="use_pose_with_covariance" value="true"/>
    <arg name="use_twist_with_covariance" value="true"/>
    <arg name="tf_rate" value="50.0"/>
    <arg name="twist_rate" value="25.0"/>
    <arg name="input_pose_name" value="dummy_pose"/>
    <arg name="input_pose_with_cov_name" value="$(var out_ndt_pose_with_covariance)"/>
    <arg name="input_twist_name" value="dummy_twist"/>
    <arg name="input_twist_with_cov_name" value="$(var out_twist_with_covariance)"/>
    <arg name="output_pose_name" value="pose"/>
    <arg name="output_odom_name" value="$(var in_ekf_odom)"/>
    <arg name="output_pose_without_yawbias_name" value="/localization/deviation_evaluator/ekf_localizer/pose_without_yawbias"/>
    <arg name="output_pose_with_covariance_without_yawbias_name" value="/localization/deviation_evaluator/ekf_localizer/pose_with_covariance_without_yawbias"/>
    <arg name="output_twist_name" value="twist"/>
    <arg name="output_twist_with_covariance_name" value="twist_with_covariance"/>

    <arg name="twist_stddev_wz" value="0.003 "/>

    <arg name="proc_stddev_vx_c" value="5.0"/>
    <arg name="proc_stddev_wz_c" value="1.0"/>
  </include>

  <!-- Map -->
  <group>
    <include file="$(find-pkg-share map_launch)/launch/map.launch.py" if="$(var rviz)">
      <arg name="lanelet2_map_path" value="$(var map_path)/$(var lanelet2_map_file)" />
      <arg name="pointcloud_map_path" value="$(var map_path)/$(var pointcloud_map_file)"/>
    </include>
  </group>

  <!-- Rviz -->
  <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen" args="-d $(find-pkg-share deviation_evaluator)/rviz/deviation_evaluator.rviz -s $(find-pkg-share autoware_launch)/rviz/image/autoware.png" if="$(var rviz)">
  </node>

  <executable cmd="ros2 bag record -o $(var save_dir)/ros2bag $(var out_twist_with_covariance) $(var in_ekf_gt_odom) $(var in_ekf_odom) $(var out_ndt_pose_with_covariance) $(var in_ndt_pose_with_covariance)" output="screen" />
</launch>
